zinc-k8s-operator/src/charm.py

```diff
@@ -0,0 +1,145 @@
+
+
+class ZincCharm(ops.CharmBase):
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        self._observe_events(framework)
+        self._pebble = self.unit.get_container("zinc") # For managing the workload container
+        self._zinc_config = {
+            "port": 4080,
+            "admin_user": "admin",
+            "admin_password": secrets.token_urlsafe(24),
+        }
+        self._zinc = zinc.ZincAPI(self._zinc_config["port"]) # For interacting with Zinc over HTTP
+
+    def _observe_events(self, framework: ops.Framework):
+        framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
+        framework.observe(self.on["zinc"].pebble_ready, self._on_zinc_pebble_ready)
+        framework.observe(self.on.get_admin_password_action, self._on_get_admin_password)
+
+    def _on_collect_unit_status(self, event: ops.CollectStatusEvent):
+        """Report the status of the unit."""
+        event.add_status(self._get_unit_status_for_container())
+        # For a more complex charm, there might be other aspects to check.
+
+    def _get_unit_status_for_container(self) -> ops.StatusBase:
+        """Determine the status of the workload container and return an appropriate unit status."""
+        try:
+            service_info = self._pebble.get_service("zinc")
+        except (ops.pebble.APIError, ops.ModelError):
+            return ops.MaintenanceStatus("waiting for Pebble to be ready")
+        if not service_info.is_running():
+            return ops.MaintenanceStatus("waiting for Zinc to start")
+        # Zinc didn't exit on startup, but it might not be ready yet, so check its health.
+        checks = self._pebble.get_checks(level=ops.pebble.CheckLevel.READY)
+        for check_info in checks.values():
+            if check_info.status == ops.pebble.CheckStatus.DOWN:
+                return ops.MaintenanceStatus("waiting for Zinc to be ready")
+                # TODO: Should we return BlockedStatus instead?
+                #       To inspect failing checks, the admin could run:
+                #       ```
+                #       juju ssh <unit> PEBBLE_SOCKET=/charm/containers/zinc/pebble.socket \
+                #       /charm/bin/pebble checks --level ready
+                #       ```
+                #       The command needs to be like this because the OCI image for Zinc has
+                #       no shell binaries (so `juju ssh --container zinc <unit> ...` won't work).
+        return ops.ActiveStatus()
+
+    def _on_zinc_pebble_ready(self, event: ops.PebbleReadyEvent):
+        """Use Pebble to start Zinc."""
+        self.unit.status = ops.MaintenanceStatus("starting Zinc")
+        if not self._update_zinc_config_from_peer_relation():
+            if self.unit.is_leader():
+                self._update_peer_relation_from_zinc_config()
+            else:
+                event.defer() # Wait for the leader to put Zinc config in the peer relation
```

Comment:
@tonyandrewmeyer do you think defer is desired here, or should we just listen to relation-changed on the peer relation and expect that event?

Reply:
I would do the latter, yes.